compiler/rustc_mir/src/interpret/validity.rs

```diff
@@ -522,17 +527,25 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
                 let value = self.read_scalar(value)?;
                 // NOTE: Keep this in sync with the array optimization for int/float
                 // types below!
-                if self.ctfe_mode.is_some() {
+                if !self.allow_uninit_and_ptr_numbers() {
+                    let ty_name = match ty.kind() {
+                        ty::Float(_) => "float",
+                        ty::Int(_) | ty::Uint(_) => "integer",
+                        _ => unreachable!(),
+                    };
+                    let value = try_validation!(
+                        value.check_init(),
+                        self.path,
+                        err_ub!(InvalidUninitBytes(None)) =>
+                            { "{}", value } expected { "initialized {} bytes", ty_name }
+                    );
                     // Integers/floats in CTFE: Must be scalar bits, pointers are dangerous
-                    let is_bits = value.check_init().map_or(false, |v| v.try_to_int().is_ok());
+                    let is_bits = value.try_to_int().is_ok();
                     if !is_bits {
                         throw_validation_failure!(self.path,
                             { "{}", value } expected { "initialized plain (non-pointer) bytes" }
```

Comment:
What are you trying to achieve by separating those two error messages here?

There is no such thing as "float bytes", so I am not sure what the first error even means in those cases.

Reply:
> There is no such thing as "float bytes", so I am not sure what the first error even means in those cases.

I think I wanted to indicate to the user what type Miri was expecting there, but I'm happy to change the wording to make it more accurate.

> What are you trying to achieve by separating those two error messages here?

I think I may have done it because I couldn't get the error to be reported, so I tried different changes to see if they would make the error work. I'll try to merge them again once I get the error working (that way I don't confuse myself as I'm trying to figure out the issue with the error).

I haven't made a change to the Miri engine before, so just let me know if there's anything in this PR you'd like me to change :)